clientv3/balancer.go

```diff
@@ -219,16 +240,123 @@ func (b *simpleBalancer) updateNotifyLoop() {
 	}
 }
 
+func (b *simpleBalancer) updateFailed(timeout time.Duration) {
+	if timeout < 3*time.Second {
+		timeout = 3 * time.Second
+	}
+	for {
+		select {
+		case <-time.After(timeout):
+			b.mu.Lock()
+			for k, v := range b.failed {
+				if time.Since(v) > timeout {
+					delete(b.failed, k)
+				}
+			}
+			b.mu.Unlock()
+		case <-b.stopc:
+			return
+		}
+	}
+}
+
+type addrConn struct {
+	addr   grpc.Address
+	failed time.Time
+}
+
 func (b *simpleBalancer) notifyAddrs() {
 	b.mu.RLock()
-	addrs := b.addrs
+	var addrs []grpc.Address
+	if len(b.addrs) == 1 ||
+		len(b.failed) == 0 ||
+		b.healthCheck == nil { // single-endpoint, or no failure
+		addrs = b.addrs
+	} else if len(b.failed) > 0 { // last pinned failing temporarily
+		all := make(map[string]grpc.Address)
+		for _, addr := range b.addrs {
+			all[addr.Addr] = addr
+			if _, ok := b.failed[addr.Addr]; ok {
+				continue
+			}
+			ok, err := b.healthCheck(addr.Addr)
+			if ok && err == nil {
+				addrs = append(addrs, addr)
+			}
+		}
+		if len(addrs) == 0 { // no better alternative found
+			addrs = b.addrs
+		} else { // sort that latest failed be at the end
+			addrConns := make([]addrConn, 0, len(b.failed))
```

Comment:
does grpc make any guarantees about the ordering? my understanding is it can try all the connections at once

Reply:
Right, there's no ordering and gRPC tries all at once. Still, goroutine starts in order.
Anyway, this wasn't necessary.